mirrord/layer/src/tcp_steal.rs

```diff
@@ -117,19 +155,133 @@ impl TcpHandler for TcpStealHandler {
+                    Some(Ok(bytes)) => Some(ClientMessage::TcpSteal(LayerTcpSteal::Data(TcpData {
+                        connection_id,
+                        bytes: bytes.to_vec(),
+                    }))),
+                    Some(Err(err)) => {
+                        error!("connection id {connection_id:?} read error: {err:?}");
+                        None
+                    }
+                    None => Some(ClientMessage::TcpSteal(
+                        LayerTcpSteal::ConnectionUnsubscribe(connection_id),
+                    )),
+                }
+            }
+            Some(res) = self.http_response_receiver.recv() => {
+                Some(ClientMessage::TcpSteal(LayerTcpSteal::HttpResponse(res)))
             }
-            None => Some(ClientMessage::TcpSteal(
-                LayerTcpSteal::ConnectionUnsubscribe(connection_id),
-            )),
         }
     }
+
+    /// Send a filtered HTTP request to the application in the appropriate port.
+    /// If this is the first filtered HTTP from its connection to arrive at this layer, a new
+    /// connection will be started for it, otherwise it will be sent in the existing connection.
+    pub(crate) async fn forward_request(&mut self, request: HttpRequest) -> Result<(), LayerError> {
+        if let Some(sender) = self.http_request_senders.get(&request.connection_id) {
+            sender
+                .send(request)
+                .await
+                .map_err::<HttpForwarderError, _>(From::from)?
+        } else {
+            self.create_http_connection(request).await?
+        }
+        Ok(())
+    }
+
+    /// Manage a single tcp connection, forward requests, wait for responses, send responses back.
+    async fn connection_task(
+        mut request_receiver: Receiver<HttpRequest>,
+        mut sender: SendRequest<Full<Bytes>>,
+        response_sender: Sender<HttpResponse>,
+        port: Port,
+        connection_id: ConnectionId,
+    ) -> Result<(), HttpForwarderError> {
+        // Listen for more requests in this connection and forward them to app.
+        while let Some(req) = request_receiver.recv().await {
+            let request_id = req.request_id;
+            // Send to application.
+            let res = sender.send_request(req.request.into()).await?;
+            let res =
+                HttpResponse::from_hyper_response(res, port, connection_id, request_id).await?;
+            // Send response back to forwarder.
+            response_sender.send(res).await?;
+        }
+        Ok(())
+    }
+
+    /// Create a new TCP connection with the application to send all the filtered HTTP requests
+    /// from this connection in.
+    /// Spawn a task that receives requests on a channel and sends them to the application on that
+    /// new TCP connection. The sender of that channel is stored in [`self.request_senders`].
+    /// The requests from the tests will arrive together with responses from all other connections
```

Comment:
What tests?

Reply:
Meant tasks, but now reworded this line.